### DISCUSSION TITLE
Transfer of variable-dependent property between multiapps

### DISCUSSION THREAD

**Discussion by rh201** [idaholab/moose#18790] [Q&A General] [answered]
Hi Moosers,

I'm trying to pass a material property (K) from the main app to the sub app  using multiapps, as code below in main app:

```
[MultiApps]
  [sub]
    type = TransientMultiApp
    input_files = sub.i
    cli_args = 'K=${K}'
    execute_on = 'TIMESTEP_END'
  []
[]
```

and define K as a GenericMaterial in sub.
 When K is a constant, it works well. Now I want to make it variable (temperature) -dependent, but it fails. 
Then I tried to define it as Auxvariable and use [transfer] block

```
[to_K]
     type = MultiAppCopyTransfer
     multi_app = sub
     direction = to_multiapp
     variable = K
     source_variable = K
   []
```
and retrieve it with MaterialRealAux, but it also don't work. Is this the right way to solve the problem?  How can I transfer this kind of nonlinear variable-dependent material property to sub app? Thanks.

Any comments or suggestions are highly appreciated.

**Comment by GiudGiud** [comment, marked as answer]
Hello

The first block you pasted gives a command line argument to the subapp. More than that needs to be transferred. The whole field of material property values needs to make its way over. 

The second block is the way to go. What do you mean by it doesnt work? Does `K` exist as a variable on both apps? Are they both populated with the right values?

Guillaume

**Reply by rh201** [reply]
@GiudGiud Thanks for replying. 

Yes, K(T) is used in both apps, and T is calculated in main, after K is updated I want it to be passed to sub. 

I define K as Auxvariable and Auxkernel in both apps

```
[AuxVariables]
  [K]
  []
[]
[AuxKernels]
  [K]
    type = ADMaterialRealAux
    variable = K
    property = K
  []
[]
```

and in main app, define K as material property

```
  [K]
    type = ADPiecewiseLinearInterpolationMaterial
    variable = 'T'
    property = 'K'
    x = '300    600'
    y = '42.47  21.235'
    outputs = exodus
  []
```

The error message is 
> *** ERROR ***                                                                                              
 The following error occurred in the object "K", of type "ADMaterialRealAux".          
 Nodal AuxKernel 'K' attempted to reference material property 'property'                  
 Consider using an elemental auxiliary variable for 'K'. 
 
What may be the problems? and how can I retrieve K in the sub in the right way? Thanks

**Reply by GiudGiud** [reply]
looks like an issue with the type of the variable. 
Can you make variable K a CONST MONOMIAL?  family=MONOMIAL, order = CONST

please use the triple quotes to format your posts

**Reply by rh201** [reply]
Hi, I changed K to "  order = CONSTANT    family = MONOMIAL" , and  has error below

> *** ERROR ***                                                                                              
The following error occurred in the object "MOOSE Problem", of type "FEProblem".                           
Material property 'K', requested by 'K' is not defined on block 0 

(used the triple quotes but it doesn't show as code format :(

**Reply by GiudGiud** [reply]
great we came through the first error. 
ok so now, is there multiple subdomains in your simulation

Also can you please rename everything? Like everything is called K and it s really confusing.
How about K_var, K_material, K_matprop, K_auxkernel for the names

**Reply by rh201** [reply]
@GiudGiud Thank you for your patience.

Great advice! Now the error message is more clear:

> *** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".
Material property 'K', requested by 'K_AuxKer_Sub' is not defined on block 0

Now the question seems down to how to convert the AuxVariable K to a material property in sub app.

**Reply by GiudGiud** [reply]
oh this is from the subapp
Ok the issue is that the material property is not defined in the subapp. It s only defined in the main app. 
So you should use a Transfer to copy the auxvariable from the main app to the subapp
See this
https://mooseframework.inl.gov/syntax/Transfers/index.html#!

Then you can work on turning the aux variable into a material property.

**Reply by rh201** [reply]
Hi, as you can see from my initial question, I have the [Transfer] block to transfer K from main to sub already.

**Reply by GiudGiud** [reply]
Ok but have you created the material property in the subapp?

**Reply by GiudGiud** [reply]
actually you need some material that will have the same value as a variable. There are a few of those in Navier Stokes, but in the framework I m not sure?
If you cant find one, then I suggest you use the VarCouplingMaterial test object,
Test objects may be used by passing the '--allow-test-objects' flag on the command line

Set the "var" argument to the K variable in the subapp

**Reply by fdkong** [reply]
> Now the question seems down to how to convert the AuxVariable K to a material property in sub app.

Something like this can covert an AuxVariable/NonlinearVariable  to a material property 


```
[AuxVariables]
  [vt]
  []
[]
```


```
[Materials]
  [diff]
    type = ParsedMaterial
    f_name = D
    args = 'vt'
    function = 'vt'
  []
[]
```

**Reply by GiudGiud** [reply]
oh cant believe I didnt know that

**Reply by rh201** [reply]
@fdkong @GiudGiud Thanks for your help! ParsedMaterial works for the problem.
